test/queue.f90

```diff
@@ -0,0 +1,171 @@
+integer                            :: capacity
+
+allocate( queue )
+allocate( queue%data(1:capacity) )
+
+queue%full  = .false.
+queue%start = 1
+queue%end   = 0
+end subroutine queue_create
+
+! queue_destroy --
+!     Destroy a queue
+! Arguments:
+!     queue       Pointer to the queue to be destroyed
+! Note:
+!     This version assumes that there are no
+!     pointers within the data that need deallocation
+!
+subroutine queue_destroy( queue )
+type(QUEUE_STRUCT), pointer  :: queue
+
+deallocate( queue%data )
+deallocate( queue )
+end subroutine queue_destroy
+
+! queue_empty --
+!     Check if the queue is empty
+! Arguments:
+!     queue       Pointer to the queue
+! Result:
+!     logical indicating if the queue is
+!     empty or not
+!
+logical function queue_empty( queue )
+type(QUEUE_STRUCT), intent(in)  :: queue
+
+queue_empty = .not. queue%full .and. &
+    queue%end .eq. queue%start - 1
+
+end function queue_empty
+
+! queue_full --
+!     Check if the queue is full
+! Arguments:
+!     queue       Pointer to the queue
+! Result:
+!     logical indicating if the queue is
+!     full or not
+!
+logical function queue_full( queue )
+type(QUEUE_STRUCT), intent(in)  :: queue
+
+queue_full = queue%full
+
+end function queue_full
+
+! queue_start_data
+!     Return the data stored at the start,
+!     but leave them in
+! Arguments:
+!     queue      Queue to be examined
+! Result:
+!     Data stored at the start
+! Note:
+!     With an empty queue, random data
+!     are returned!
+!
+function queue_start_data( queue ) result(data)
+type(QUEUE_STRUCT), intent(in) :: queue
+integer                        :: data
+
+data = queue%data(queue%start)
+
+end function queue_start_data
+
+! queue_retrieve_data
+!     Return the data stored at the top,
+!     and remove them from the queue
+! Arguments:
+!     queue      Queue to be examined
+! Result:
+!     Data stored at the top, afterwards
+!     removed
+! Note:
+!     With an empty queue, random data
+!     are returned!
+!
+function queue_retrieve_data( queue ) result(data)
+type(QUEUE_STRUCT)             :: queue
+integer                        :: data
+
+data = queue%data(queue%start)
+
+if ( .not. queue_empty(queue) ) then
+    queue%start = queue%start + 1
+    if ( queue%start .gt. size(queue%data) ) then
+        queue%start = 1
+    endif
+    queue%full = .false.
+endif
+
+end function queue_retrieve_data
+
+! queue_append_data
+!     Append data to the end of the queue
+! Arguments:
+!     queue      Queueu to which to add the data
+!     data       The data to be added
+!     success    Indicates success or not
+!
+subroutine queue_append_data( queue, data, success )
+type(QUEUE_STRUCT)           :: queue
+integer, intent(in) :: data
+logical, intent(out)         :: success
+
+success = .not. queue_full( queue )
+if ( success ) then
+    queue%end = queue%end + 1
+    if ( queue%end .gt. size(queue%data) ) then
+        queue%end = 1
```

Comment:
May be make queue full

Reply:
There is a check to make the queue full later in the function. The reason that is check exists is because the queue is circular. Meaning when we add and remove from the array, they aren't being added to the beginning or end. So this check allows the queue to wrap around and make sure that the full array is utilized.